cpp/modules/deck.gl/layers/src/scatterplot-layer/scatterplot-layer.cpp

```diff
@@ -158,38 +135,21 @@ void ScatterplotLayer::updateState(const Layer::ChangeFlags& changeFlags, const
 
 void ScatterplotLayer::finalizeState() {}
 
-void ScatterplotLayer::drawState(wgpu::RenderPassEncoder pass) {  // uniforms
-  /*
-    const {viewport} = this->context;
-    const {
-      radiusScale,
-      radiusMinPixels,
-      radiusMaxPixels,
-      stroked,
-      filled,
-      lineWidthUnits,
-      lineWidthScale,
-      lineWidthMinPixels,
-      lineWidthMaxPixels
-    } = this->props;
-
-    const widthMultiplier = lineWidthUnits === 'pixels' ?
-      viewport.metersPerPixel : 1;
-
-    this->state.model
-      .setUniforms(uniforms)
-      .setUniforms({
-        stroked: stroked ? 1 : 0,
-        filled,
-        radiusScale,
-        radiusMinPixels,
-        radiusMaxPixels,
-        lineWidthScale: lineWidthScale * widthMultiplier,
-        lineWidthMinPixels,
-        lineWidthMaxPixels
-      })
-      .draw();
-    */
+void ScatterplotLayer::drawState(wgpu::RenderPassEncoder pass) {
+  auto props = std::dynamic_pointer_cast<ScatterplotLayer::Props>(this->props());
+  float widthMultiplier = props->lineWidthUnits == "pixels" ? this->context->viewport->metersPerPixel() : 1.0;
+  auto lineWidthScale = props->lineWidthScale * widthMultiplier;
+
+  ScatterplotLayerUniforms layerUniforms{
+      props->opacity, props->radiusScale,        props->radiusMinPixels,    props->radiusMaxPixels,
+      lineWidthScale, props->lineWidthMinPixels, props->lineWidthMaxPixels, props->stroked ? 1.0f : 0.0f,
+      props->filled};
+  for (auto const& model : this->getModels()) {
+    // Layer uniforms are currently bound to index 1
+    model->setUniforms(
+        std::make_shared<garrow::Array>(this->context->device, &layerUniforms, 1, wgpu::BufferUsage::Uniform), 1);
```

Comment:
Moved buffer creation, data setting and binding across `initialize`/`update`/`draw` state methods.

We now create a single buffer that gets filled with data, which works well except that `updateState` isn't called when viewport updates (is that supposed to be the case?), so `widthMultiplier` isn't correct.

Temporarily made it so that `updateState` is called from `drawState`, which can easily be removed once diffing/dirty flags work correctly

Reply:
> We now create a single buffer that gets filled with data, which works well except that updateState isn't called when viewport updates (is that supposed to be the case?), so widthMultiplier isn't correct.

I suggest we design this somewhat independently from JS. The JS API didn't moved all uniform updating into updateState. It also implements the "uniform animation" feature in a JavaScript dependent way. 

Updating when viewport changes can trigger a lot of updates, e.g. when using multiple views we get multiple updates every frame. So we want to minimize calls if possible.

Reply:
Not sure I fully understand where we should be doing the updates then, let's sync on this one